emergence_lib/src/signals/equation.rs

```diff
@@ -0,0 +1,163 @@
+            //       equivalent to world distance.
+            let x = source.distance_to(at_tile.hex) as f32;
+            sum += strength / (4.0 * (1.0 + DECAY_RATE * time) * consts::PI * DIFFUSIVITY * time)
+                * (-0.25 * x * x / (DIFFUSIVITY * time)).exp();
+            // Same expression, with precision loss optimized by Herbie (though it should be negligible):
+            // sum += 0.25 * consts::FRAC_1_PI * (strength / DIFFUSIVITY)
+            //     * ((x / DIFFUSIVITY * (-0.25 * x / time)).exp() / time);
+        }
+        SignalStrength(sum)
+    }
+
+    /// Start emitting a signal of given `strength` at the `source_tile` position.
+    #[inline]
+    pub fn emit_signal(&mut self, source: TilePos, SignalStrength(strength): SignalStrength) {
+        self.emissions.push_back(SignalEmission {
+            time: self.time,
+            strength,
+            source,
+        });
+    }
+
+    /// Sets the time at which the equation will be evaluated to the given `current_time`. This
+    // doesn't recompute anything and is very cheap.
+    #[inline]
+    pub fn advance_time(&mut self, current_time: f32) {
+        debug_assert!(self.time <= current_time);
+        self.time = current_time;
+    }
+
+    /// Get rid of signal emissions that do contribute to negligible amounts to the global
+    /// signal.
+    ///
+    /// If `true` is returned, any further evaluation of this equation will be
+    /// [`SignalStrength::ZERO`][super::SignalStrength::ZERO] unless
+    /// [`emit_signal`][DiffusionEquation::emit_signal] is called.
+    pub fn trim_neglible_emissions(&mut self) -> bool {
+        trace!("Attempting trimming among {} sources", self.emissions.len());
+        let mut visited = 0;
+        while visited < self.emissions.len() {
+            match self.emissions.pop_back() {
+                // If the emission has been there long enough, decay will eventually take over.
+                // The comparison here checks whether the area under the curve of the solution
+                // (which could be seen as the amount of pheromones) is lower than some threshold
+                // proportional to the strength of the signal. Since this area is directly
+                // proportional to the strength as well, this implies the signal emission will
+                // be trimmed after some time inversely proportional to the decay rate.
+                // Intuitively, this would be like waiting until the most of pheromones
+                // dispersed by the emitted signal decay, the precise amount we consider
+                // negligible being more for fast decaying signals (we'll trim it earlier), and
+                // less for slowly decaying ones (we'll trim it later)
+                Some(SignalEmission { time, strength, .. })
+                    if DECAY_THRESHOLD * (1.0 + DECAY_RATE * (self.time - time)) > strength => {}
```

Comment:
I'm having trouble following the math here when skimming; I think intermediate variables will make this clearer.